Random Double Battle: Update level scales based on item/ability

ascriptmaster opened the pull request:
This will prevent Pokemon like Azumarill from being too powerful due to
their low base stats with an ability that boosts it to incredible levels.

Open for discussion: Do we need to scale for these items/abilities:
- Sheer Force
- -ate abilities
- Expert Belt

Please discuss a lot before merging!

ascriptmaster commented:
@TheImmortal @Joimer

Joimer commented:
Sheer force, -ate: yes. Other permanent boost abilities that make a Pokémon good should be taken into account too.
Items in general: Better not, besides Soul Dew (which isn't used). In fact, can we get rid of Life Orb check and just leave in Eviolite check (as it is damn powerful)?

ascriptmaster commented:
What about Thick Club?

Joimer commented:
Thick Club and Light Ball make useless Pokémon slightly less useless, I don't think they need a level nerf since they aren't exactly OP to begin with.

ascriptmaster commented:
And how should Sheer Force be scaled? Should it do its stat calculation by factoring in `counter['sheerforce']/damagingMoves.length` or just the naive boost? Or maybe some percentage in between?

Joimer commented:
If we account for sheer force, it shouldn't be much. Any mon with sheer force will have a boosted attack regardless, so we don't need to count them. Naive boost will do.

ascriptmaster commented:
Then what of -ate?

Joimer commented:
Same drill.

ascriptmaster commented:
Added most boosting abilities (Hustle, Mega Launcher, Protean, Iron Fist...) and removed all item checks other than Eviolite.

How should I deal with Multiscale? Pokemon with it effectively have higher HP than usual, but I'm not sure how I'd scale it.

Joimer commented:
Multiscale is broken by a first hit, priority, double target, etc, dragonite already has decent level without nerf I think.

Slayer95 commented:
Do you think that BSR's usefulness could apply to Doubles, too? http://www.smogon.com/forums/threads/stats-ratings.39556/

If it were, then maybe we could just plug the modified stats into X-Act's formula?

Joimer commented:
I don't know if we would be adding too much code for level balance then tbh

ascriptmaster commented:
That seems like something more tailored towards EV generation than this, I think. Randdubs uses its own EV generator instead of going 84-flat, after all.

ascriptmaster commented:
Yeah, I don't want to bother with that level of calculations. We'd need to build a lookup table to get speed factors and run a bunch of other calculations that aren't worth it for the situation.

TheImmortal commented on data/scripts.js:
```diff
@@ -2969,8 +2969,31 @@ exports.BattleScripts = {
+		var templateAbility = (baseTemplate === template ? ability : template.abilities[0]);
+		if (templateAbility === 'Huge Power' || templateAbility === 'Pure Power') {
+			bst += (evs.atk > evs.spa ? template.baseStats.atk : template.baseStats.spa) + 45;
+		} else if (templateAbility in {'Guts':1, 'Parental Bond':1, 'Hustle':1, 'Mega Launcher':1, 'Protean':1}) {
```
Guts and Hustle has negative effects, losing health and inaccuracy respectively, so I wouldn't put that on the same level as those.

Joimer replied:
Agreed.

TheImmortal commented:
Actually, after thinking about it, I don't think you need to add every single ability/item. The major ones that double stats with no side effect (Pure Power / Fur Coat) should be included but everything else shouldn't imo.

Not sure what @Joimer would say about that.

Joimer commented:
Indeed, we only shall take into account the abilities which directly enhace Pokémon with no drawbacks. I'd still lower somewhat some -ate users, because, for instance, Sylveon's curren't level can lead to problems with Hyper Voice spam.

ascriptmaster commented:
I "nerfed" the BST boost Guts/Hustle give to 1.2 instead of 1.5, but would you rather I remove it entirely?

TheImmortal commented:
Are you really going to add every ability? Speed Boost improves a Pokemon significantly, for example. I suggest remove all these and just include broken ones like Huge Power and Parental Bond.

TheImmortal commented on data/scripts.js:
```diff
@@ -2969,8 +2969,32 @@ exports.BattleScripts = {
+			bst -= 0.5 * (template.baseStats.atk + template.baseStats.spe) + 20;
+		}
+		if (item === 'Eviolite') {
+			bst += 0.5 * (template.baseStats.def + template.baseStats.spd) + 10;
```
Why are you adding 10?

TheImmortal commented on data/scripts.js:
```diff
@@ -2969,8 +2969,32 @@ exports.BattleScripts = {
+		// Cut some ability cases a little slack since many of these mons can't have any other kind of boosting item (based on holistic judgment)
+		var templateAbility = (baseTemplate === template ? ability : template.abilities[0]);
+		if (templateAbility === 'Huge Power' || templateAbility === 'Pure Power') {
+			bst += template.baseStats.atk + 45;
```
Why not just double the attack stat?

ascriptmaster replied:
You mean, before the BST total is calculated?

TheImmortal replied:
Oh hmm, I didn't realize that. Then I change my question to why are you adding 45?

ascriptmaster replied:
As mentioned in the comments, if you increase a stat by `M`, then it increases the effective base stat by `M * [normalized base stat] = M * ([base stat]+50)`. In this case, since Attack goes up by an equal amount, `M=1`. Using 45 instead of 50 is holistic judgment.

TheImmortal commented on data/scripts.js:
```diff
@@ -2969,8 +2969,32 @@ exports.BattleScripts = {
+		var templateAbility = (baseTemplate === template ? ability : template.abilities[0]);
+		if (templateAbility === 'Huge Power' || templateAbility === 'Pure Power') {
+			bst += template.baseStats.atk + 45;
+		} else if (templateAbility in {'Parental Bond':1, 'Mega Launcher':1, 'Protean':1}) {
```
For Parental Bond, I guess you can just increase the attack stat by 50%?

TheImmortal commented on data/scripts.js:
```diff
@@ -2969,8 +2969,32 @@ exports.BattleScripts = {
+			// Guts and Hustle are in this group instead of the 1.5x group because they have drawbacks in their abilities
+			bst += 0.2 * template.basestats.atk + 10;
+		} else if (templateAbility === 'Fur Coat') {
+			bst += template.baseStats.def + 13;
```
Why not just double it?

Joimer commented:
Question: What levels are we getting right now on the important mons with the current state of this balance code? @ascriptmaster

ascriptmaster commented:
- Azumarill: 87 -> 78
- Diggersby tho?: 87 -> 77
- Medicham: 88 -> 78
- Mega Med: 78 -> 70
- Mega Maw: 81 -> 70
- Sylveon: 77 -> 72
- Aurorus: 77 -> 73

Mega Kang/Alt/Mence/Garde/Pinsir all stay at level 70.

I'll calc the rest later (lots of Sheer Force mons) but what do you think of these?

ascriptmaster commented:
I removed Iron First/Hustle/Guts ability check.

In other news, Chansey's level goes from 84 to 78 under these changes.

ascriptmaster commented:
TI wanted me to remove Adapatability, Tough Claws and -ate Abilities, so I did. Sylveon being Level 77 should be fine even with Hyper Voice spam.

TheImmortal commented:
I also think you should remove Sheer Force. Finally, Protean should maybe get half the boost of Parental Bond.

ascriptmaster commented:
I removed Sheer Force and put Protean in its old place (so Protean is treated as if it gives a 1.3x boost)

I also bumped Huge Power and Pure Power's level scales to its actual effective base stat value. Fur Coat / Eviolite still have reduced-ish base stat increases (+13 instead of +18)

TheImmortal commented:
Hmm, how about slightly lowering Eviolite (increasing level) because it loses out on the item benefit that everything else has, so for example it would be the same level as another Pokemon with the same defenses after Eviolite (assuming other stats are equal...) but the other Pokemon has the benefit of an item.

TheImmortal commented:
Actually, just remove the added number. And the same for the abilities because of the same reason that Pokemon benefit from their ability, etc...

ascriptmaster commented:
The added constant numbers are removed.
